java/ddlogapi.c

```diff
@@ -203,24 +291,30 @@ void commit_dump_callback(void* callbackInfo, table_id tableid, const ddlog_reco
         env, cbi->obj, cbi->method, (jint)tableid, (jlong)rec, (jboolean)polarity);
 }
 
-JNIEXPORT jint JNICALL Java_ddlogapi_DDlogAPI_ddlog_1transaction_1commit_1dump_1changes(
+JNIEXPORT void JNICALL Java_ddlogapi_DDlogAPI_ddlog_1transaction_1commit_1dump_1changes(
     JNIEnv * env, jobject obj, jlong handle, jstring callback) {
 
-    if (callback == NULL)
-        return ddlog_transaction_commit_dump_changes((ddlog_prog)handle, NULL, 0);
+    if (callback == NULL) {
+        if (ddlog_transaction_commit_dump_changes((ddlog_prog)handle, NULL, 0) < 0) {
+            throwDDlogException(env);
+            return;
+        }
+    }
 
     struct CallbackInfo* cbinfo = createCallback(env, obj, callback, "(IJZ)V");
     if (cbinfo == NULL)
-        return -1;
+        return;
 
-    int result = ddlog_transaction_commit_dump_changes((ddlog_prog)handle,
-                                                       commit_dump_callback,
-                                                       (uintptr_t)cbinfo);
+    if (ddlog_transaction_commit_dump_changes(
+                (ddlog_prog)handle,
+                commit_dump_callback,
+                (uintptr_t)cbinfo) < 0) {
+        throwDDlogException(env);
```

Comment:
How about a helper function check(env, exitCode) which throws when the exitCode is negative?

Reply:
I had that initially, but it's a bit tricky, cause we still need to check the error code and act on it, whether we raise the exception or not.  It was not very ergonomic.